File: elixir/bob/bob.exs
defmodule Teenager do
  def hey(message) do
    message = String.strip(message || "")

    cond do
      silent?(message) ->
        "Fine. Be that way!"
      shouting?(message) ->
        "Woah, chill out!"
      asking?(message) ->
        "Sure."
      true ->
        "Whatever."
    end
  end

  defp shouting?(message) do
    message == String.upcase(message)
  end

  defp asking?(message) do
    String.ends_with?(message, "?")
  end

  defp silent?(message) do
    message == ""
  end
end

File: scala/bob/src/main/scala/bob.scala
class Bob {
  def hey(msg: String) = msg match {
    case m if isSilent(m) =>
      "Fine. Be that way!"
    case m if isShouting(m) =>
      "Woah, chill out!"
    case m if isAsking(m) =>
      "Sure."
    case _ =>
      "Whatever."
  }

  private def isSilent(msg: String) =
    msg.trim.isEmpty

  private def isShouting(msg: String) = {
    val loud  = msg.toUpperCase
    val quiet = msg.toLowerCase

    msg == loud && loud != quiet
  }

  private def isAsking(msg: String) =
    msg.endsWith("?")
}
